src/mewbot/io/file_storage.py

```diff
@@ -0,0 +1,162 @@
+    def consumes_outputs() -> Set[Type[OutputEvent]]:
+        """
+        Defines the set of output events that this Output class can consume
+        :return:
+        """
+        return {
+            CreateDirectoryOutputEvent,
+            WriteToFileOutputEvent,
+            DeleteFileOutputEvent,
+        }
+
+    async def output(self, event: OutputEvent) -> bool:
+        """
+        Does the work of transmitting the event to the world.
+        :param event:
+        :return:
+        """
+        if not isinstance(event, FSOutputEvent) or self._output_path is None:
+            return False
+
+        if not self._output_path_exists:
+            self._logger.warning(
+                "Cannot output - output_path '%s' does not exist", self._output_path
+            )
+            return False
+
+        path = self._output_path.joinpath(event.path).resolve()
+
+        if not path.is_relative_to(self._output_path):
+            self._logger.error(
+                "Refusing to write to %s, as it would result in a file outside of %s",
+                event.path,
+                self._output_path,
+            )
+            return False
+
+        async_path = aiopath.AsyncPath(path)
+
+        if isinstance(event, CreateDirectoryOutputEvent):
+            return await self._process_directory_create_event(async_path)
+
+        if isinstance(event, WriteToFileOutputEvent):
+            return await self._process_file_write_event(async_path, event)
+
+        if isinstance(event, DeleteFileOutputEvent):
+            return await self._process_delete_file_event(async_path)
+
+        raise NotImplementedError(f"Unexpected event type {event}")
+
+    @staticmethod
+    async def _process_directory_create_event(path: aiopath.AsyncPath) -> bool:
+        await path.mkdir(parents=True, exist_ok=True)
+        return True
+
+    async def _process_file_write_event(
+        self, path: aiopath.AsyncPath, event: WriteToFileOutputEvent
+    ) -> bool:
+        mode: str = "w"
+
+        if not event.may_overwrite:
+            mode = "x"
+        if event.append:
+            mode = "a"
+
+        full_mode = mode if isinstance(event.file_contents, str) else mode + "b"
+
+        async with path.open(mode=full_mode) as outfile:
+            await outfile.write(event.file_contents)
+
+        return True
+
+    @staticmethod
+    async def _process_delete_file_event(path: aiopath.AsyncPath) -> bool:
+        await path.unlink()
+        return True
+
+
+@dataclasses.dataclass
+class FSOutputEvent(OutputEvent):
+    """
+    Base class for generic input events generated by the file system.
+    """
+
+    path: Union[str, os.PathLike[str]]
+
+
+@dataclasses.dataclass
+class CreateDirectoryOutputEvent(FSOutputEvent):
+    """Creates a directory at the given path (and any parent directories)"""
+
+
+@dataclasses.dataclass
+class WriteToFileOutputEvent(FSOutputEvent):
+    """
+    Writes the given contents to the specified file.
+
```

Comment:
Suggest adding something like "within the designated _output_path" - just to emphasize that arbitary file creation is not supported.